core/runtime/src/fetch/request.rs

```diff
@@ -0,0 +1,223 @@
+//! The `Request` JavaScript class and adjacent types, implemented as [`JsRequest`].
+//!
+//! See the [Request interface documentation][mdn] for more information.
+//!
+//! [mdn]: https://developer.mozilla.org/en-US/docs/Web/API/Request
+use super::HttpRequest;
+use boa_engine::value::{Convert, TryFromJs};
+use boa_engine::{Finalize, JsData, JsObject, JsResult, JsString, JsValue, Trace, js_error};
+use boa_interop::boa_macros::boa_class;
+use either::Either;
+use std::collections::BTreeMap;
+use std::mem;
+
+fn add_headers_to_builder<'a>(
+    headers: impl Iterator<Item = (&'a JsString, &'a Convert<JsString>)>,
+    mut builder: http::request::Builder,
+) -> JsResult<http::request::Builder> {
+    for (hkey, Convert(hvalue)) in headers {
+        // Make sure key and value can be represented by regular strings.
+        // Keys also cannot have any extended characters (>128).
+        // Values cannot have unpaired surrogates.
+        let key = hkey.to_std_string().map_err(|_| {
+            js_error!(TypeError: "Request constructor: {} is an invalid header name", hkey.to_std_string_escaped())
+        })?;
+        if !key.is_ascii() {
+            return Err(
+                js_error!(TypeError: "Request constructor: {} is an invalid header name", hkey.to_std_string_escaped()),
+            );
+        }
+        let value = hvalue.to_std_string().map_err(|_| {
+            js_error!(
+                TypeError: "Request constructor: {:?} is an invalid header value",
+                hvalue.to_std_string_escaped()
+            )
+        })?;
+
+        builder = builder.header(key, value);
+    }
+
+    Ok(builder)
+}
+
+fn add_vec_headers_to_builder(
+    headers: &[(JsString, Convert<JsString>)],
+    builder: http::request::Builder,
+) -> JsResult<http::request::Builder> {
+    add_headers_to_builder(headers.iter().map(|(k, v)| (k, v)), builder)
+}
+
+fn add_btree_headers_to_builder(
+    headers: &BTreeMap<JsString, Convert<JsString>>,
+    builder: http::request::Builder,
+) -> JsResult<http::request::Builder> {
+    add_headers_to_builder(headers.iter(), builder)
+}
```

Comment:
I don't really see the benefit of having this. The underlying calls to `headers.iter()` are so simple that they should be done inline.

Reply:
Yeah they diverged at some point and I didn't finish the cleanup. Done.